Ballot report

taskboy3000 opened the pull request:
Ballot report:

- New Admin section (needs to be secured and linked to the main app somehow)
- New Voting section that contains the Ballot Reports
- Ballot activity shown for Entries, Users and IP

jmacdotorg commented on IFComp/lib/IFComp/Controller/Admin/Voting.pm:
```diff
@@ -0,0 +1,158 @@
+    $c->stash->{ available_comps } = \@available_comps;
+
+    # Get all entries for this comp
+    my @all_entries = $entry->search({comp => $current_comp->id,
```
Instead of checking is_disqualified, please just grab all entries and then grep them through `$_->is_qualified`, which checks other qualification facets as well. (e.g. `my @qualified_entries = grep { $_->is_qualified } @all_entries`

jmacdotorg commented on IFComp/lib/IFComp/Controller/Admin.pm:
```diff
@@ -0,0 +1,49 @@
+
+=cut
+
+sub index :Path :Args(0) {
```
The access to this needs some locking down. Could you please check to see if the current user possess the 'votecounter' role? before allowing access to this action? (If they do not, just redirect them to /.

Note that role support was only added in early August, so you'll have to merge in the current state of the master branch into this one to use it.

taskboy3000 replied:
My version of the schema has role.name as CHAR(8), which means it is not possible for a role to be named 'votecounter'.  Did the schema change?

jmacdotorg commented on IFComp/lib/IFComp/Controller/Admin/Voting.pm:
```diff
@@ -0,0 +1,158 @@
+
+=cut
+
+sub index :Path :Args(0) {
```
Please make this and all other /admin/foo endpoints Chained actions, which chain off the root admin action. (See IFComp::Controller::Entry for examples of chained actions.)

Doing this will let you apply role-based authorization to to root admin action and then have it automatically applied to all these actions as well.

taskboy3000 commented on IFComp/lib/IFComp/Schema/ResultSet/Entry.pm:
```diff
@@ -0,0 +1,9 @@
+package IFComp::Schema::ResultSet::Entry;
```
Yes.  I had a lot more code in there originally.  I will remove the file.

taskboy3000 commented:
I have 
I hope I fixed the indenting.
I added chained handlers for authorization (appears to be working in my sandbox)
Since role.name is 8 chars, the ballot_report look authorize users with roles 'votecoun'.

taskboy3000 commented:
Push a change so that the role sought for is "votecounter".  This will not work until the shema is updated.

jmacdotorg commented:
First of all, I see I didn't describe very well where I suggested a grep{} go, so I just popped that in myself; hope you don't mind.

Secondly, I see on running this locally that I messed up in failing to share some very crucial information with you weeks ago.

The reason the SQL statements I shared with you earlier are so long and tortured, versus simply checking the values of the entries' "average_score" and "votes_cast" and so on, is that none of the thirteen score-related columns of the entry table get their values filled in until after the judging period is over. At that point, I run a one-off script that computes these values for all the relevant entry rows, and that's their permanent values forever.

*During* the judging period, all these fields are NULL. They are not kept up-to-date as votes come in.

So... checking the values of these columns to get data about the currently-in-progress competition won't work.

May I suggest adding methods to the Entry result object that calls the SQL to compute these various values on the fly, and then having the new Vote report controller call these instead?

(Sorry...)

taskboy3000 commented:
I rather suspected something like this was happening.

The performance of those pages will be terrible without computing those
values ahead of time.
Could these be computed nightly or even hourly?

On Tue, Oct 4, 2016 at 10:33 PM, Jason McIntosh <notifications@github.com>
wrote:

> First of all, I see I didn't describe very well where I suggested a grep{}
> go, so I just popped that in myself; hope you don't mind.
>
> Secondly, I see on running this locally that I messed up in failing to
> share some very crucial information with you weeks ago.
>
> The reason the SQL statements I shared with you earlier are so long and
> tortured, versus simply checking the values of the entries' "average_score"
> and "votes_cast" and so on, is that none of the thirteen score-related
> columns of the entry table get their values filled in until after the
> judging period is over. At that point, I run a one-off script that computes
> these values for all the relevant entry rows, and that's their permanent
> values forever.
>
> *During* the judging period, all these fields are NULL. They are not kept
> up-to-date as votes come in.
>
> So... checking the values of these columns to get data about the
> currently-in-progress competition won't work.
>
> May I suggest adding methods to the Entry result object that calls the SQL
> to compute these various values on the fly, and then having the new Vote
> report controller call these instead?
>
> (Sorry...)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmacdotorg/ifcomp/pull/62#issuecomment-251567089>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAQo4UOIJV-XwS3qbxicayf4p3Wt5dlpks5qwwxngaJpZM4J6Fe7>
> .
>



-- 
Joe Johnston - taskboy.com
